lang/test/org/partiql/lang/eval/EvaluatingCompilerFromLetTests.kt

```diff
@@ -0,0 +1,204 @@
+                expectedSql = """<< {'region': `EU`, 'maxLen': 6}, {'region': `NA`, 'maxLen': 9} >>"""),
+                session = session)
+    }
+
+    @Test
+    fun `LET clause outer query has correct value` () {
+        runTestCase(
+            EvaluatorTestCase(
+                query = "SELECT X FROM (SELECT VALUE X FROM A LET 1 AS X) LET 2 AS X",
+                expectedSql = """<< {'X': 2} >>"""),
+                session = session)
+    }
+
+    // Error test cases
+    @Test
+    fun `LET clause unbound variable` () {
+        checkInputThrowingEvaluationException(
+            "SELECT X FROM A LET Y AS X",
+            session,
+            ErrorCode.EVALUATOR_BINDING_DOES_NOT_EXIST,
+            mapOf(
+                    Property.LINE_NUMBER to 1L,
+                    Property.COLUMN_NUMBER to 21L,
+                    Property.BINDING_NAME to "Y"
+            )
+        )
+    }
+
+    @Test
+    fun `LET clause inner query binding not available in outer query` () {
+        checkInputThrowingEvaluationException(
+            "SELECT X FROM (SELECT VALUE X FROM A LET 1 AS X)",
+            session,
+            ErrorCode.EVALUATOR_BINDING_DOES_NOT_EXIST,
+            mapOf(
+                    Property.LINE_NUMBER to 1L,
+                    Property.COLUMN_NUMBER to 8L,
+                    Property.BINDING_NAME to "X"
+            )
+        )
+    }
+
+    @Test
+    fun `LET clause binding in subquery not in outer LET query` () {
+        checkInputThrowingEvaluationException(
+            "SELECT Z FROM A LET (SELECT 1 FROM A LET 1 AS X) AS Y, X AS Z",
+            session,
+            ErrorCode.EVALUATOR_BINDING_DOES_NOT_EXIST,
+            mapOf(
+                    Property.LINE_NUMBER to 1L,
+                    Property.COLUMN_NUMBER to 56L,
+                    Property.BINDING_NAME to "X"
+            )
+        )
+    }
+
+    @Test
+    fun `LET clause binding referenced in HAVING not in GROUP BY`() {
+        checkInputThrowingEvaluationException(
+            "SELECT B.id FROM B LET 100 AS foo GROUP BY B.id HAVING B.id > foo",
+            session,
+            ErrorCode.EVALUATOR_VARIABLE_NOT_INCLUDED_IN_GROUP_BY,
+            mapOf(
+                    Property.LINE_NUMBER to 1L,
+                    Property.COLUMN_NUMBER to 63L,
+                    Property.BINDING_NAME to "foo"
+            )
+        )
+    }
+
+    @Test
+    fun `LET clause binding referenced in projection not in GROUP BY`() {
+        checkInputThrowingEvaluationException(
+            "SELECT foo FROM B LET 100 AS foo GROUP BY B.id",
+            session,
+            ErrorCode.EVALUATOR_VARIABLE_NOT_INCLUDED_IN_GROUP_BY,
+            mapOf(
+                    Property.LINE_NUMBER to 1L,
+                    Property.COLUMN_NUMBER to 8L,
+                    Property.BINDING_NAME to "foo"
+            )
+        )
+    }
```

Comment:
A case where `LET` has multiple bindings and one of them depends on a binding that appears later in the `LET` clause, e.g., 

```
SELECT * FROM X 
LET 1 AS X, 
    Y AS Z, 
    3 AS Y   
```

Reply:
Got it will add.